compiler/rustc_mir_transform/src/lib.rs

```diff
@@ -352,14 +353,22 @@ fn mir_promoted(
 
     // What we need to run borrowck etc.
     let promote_pass = promote_consts::PromoteTemps::default();
+    let promote_array = promote_consts_local_arrays::PromoteArraysOpt::default();
     pm::run_passes(
         tcx,
         &mut body,
-        &[&promote_pass, &simplify::SimplifyCfg::PromoteConsts, &coverage::InstrumentCoverage],
+        &[
+            &promote_pass,
+            &promote_array,
+            &simplify::SimplifyCfg::PromoteConsts,
+            &coverage::InstrumentCoverage,
+        ],
         Some(MirPhase::Analysis(AnalysisPhase::Initial)),
     );
 
-    let promoted = promote_pass.promoted_fragments.into_inner();
+    let mut promoted = promote_pass.promoted_fragments.into_inner();
+    let array_promoted = promote_array.promoted_fragments.into_inner();
+    promoted.extend(array_promoted);
```

Comment:
which does mean you won't be able to use the existing promotion scheme, but would need to start looking into `create_def` and query feeding, which is probably not ready to support this use case yet. I have not yet given it much thought what is needed to fully support that, but if you want we can look into this together.

Reply:
then again, if all we're doing is creating non-generic static items, that already has precedent (we do that for nested statics), so likely you can do the same in an optimization.

Reply:
With GVN or similar you don't even need to create new constants and MIR bodies, you can just stick the fully evaluated constant into a MIR constant